Add 2021-01-20-reproducible-progress.md

marmarek opened the pull request.

marmarek commented:
@fepitre see if I haven't missed anything, especially some other tools you have contributed to in the process.
Also, I know the qubes-rebuilder repo is still private. We can publish the article after it's switched to public.

@andrewdavidwong besides the usual language check, please see if I haven't failed to explain some not obvious technical parts. It's easy to assume something is obvious when it's your daily job.

andrewdavidwong commented on 2021-01-20-reproducible-progress.md:
```diff
@@ -0,0 +1,81 @@
+
+The next step on the journey is enabling an arbitrary user or organization to challenge the official packages --- to take the sources, together with the `buildinfo` file describing how it was built and verify that they result in the same binary package. Theoretically, there is a [`debrebuild.pl`] script in Debian's `devscripts` package that does exactly this. We started by adding missing parts to ...
+
+To ease the rebuilder setup, we are also in the process of creating a simple [rebuilder orchestrator] that monitors specified repositories and does the package build verification automatically when necessary. The goal is to give anyone who wants to verify our (or others) packages an easy way to get started. The decision of who to trust is separate, but in order to have a choice, rebuilders must e ...
```
> The decision of who to trust is separate, but in order to have a choice, rebuilders must exist in the first place.

This sentence is somewhat mysterious. I suggest adding more context explaining what you have in mind.

marmarta replied:
Maybe something like: "The goal is to give anyone who wants to verify our (or others) packages an easy way to get started. Of course, they have to choose which rebuilder to trust (and this is a separate choice) - but there must exist some rebuilders to choose from in the first place."

Although I would really welcome a definition of a rebuilder here, tbh I'm not sure what it really means.

marmarek replied:
Rebuilder is a thing that given a binary package and its source (plus some extra metadata if applicable), tries to build that package again (rebuild), to verify if that binary really comes from the source it claims to use.

andrewdavidwong reviewed (commented):
> @andrewdavidwong besides the usual language check, please see if I haven't failed to explain some not obvious technical parts. It's easy to assume something is obvious when it's your daily job.

This is a fairly technical article. It assumes that the reader is already familiar with (and interested in) CI and reproducible builds. If that's your target audience, then it may be fine as is. However, if your aim is to make the article more accessible, here are some possibilities:

 - Add an introductory section explaining the role of CI and reproducible builds in the Qubes OS Project.
 - At the start of the intro, briefly explain (or link to explanations) of what CI and reproducible builds are. This is for readers who aren't familiar with these concepts. (For example, we never actually state that "CI" is an abbreviation for "continuous integration." We should do so, but merely stating that won't be sufficient for someone who doesn't know what continuous integration is. If we can explain that briefly here, great. If not, link to an external explanation. Likewise for reproducible builds.)
 - Explain why CI is important for the Qubes OS Project and how it's mainly used. This is for people who (now) understand what CI is but who may not know exactly how the Qubes OS Project operates.
 - Explain why reproducible builds are important for Qubes. This is for anyone who isn't already aware (or perhaps has never thought about) the security implications of reproducibility, especially if there are any Qubes-specific considerations.

fepitre commented:
https://github.com/fepitre/qubes-rebuilder is now public.

marmarta commented:
I'd propose the following intro; it's fairly brief, but I hope it gives a bit more context. It can also be completely wrong technically, so correct me if I'm wrong somewhere, @marmarek 


Improvements in testing and building: GitLab CI and reproducible builds

Over the last couple of months, we had made some significant changed to two important parts of Qubes development process: testing and building. If terms like CI or reproducible builds are clear and obvious to you, you can jump to detailed technical explanations below - this introduction will give a little more context to those of our readers who may be encountering those terms for the first time.

Automated testing is a huge part of the software development process. It may not be very visible on the outside, but in practice, it is crucial and spares developers many, many hours of manual tests that would still often miss bugs and problems. In Qubes development, we're using an approach called Continuous Integration (CI) - local changes made by the developers are uploaded and tested remotely, using dedicated CI solutions. This is very important for both maintaining consistent code quality (all changes will be tested) and for making development easier for the developers (testing Qubes is not easy - it's a whole operating system, and doing the testing on the same system you're developing in is a bit like building a rocket landing system en route to Mars... perhaps not impossible, but very stressful).

The second part we made improvements in is the build process. The idea of 'reproducible builds' may sound quite silly at first - it just means that the same source code results in exactly the same executable program, and what exactly would be difficult here? After all, computers are not random - right? Unfortunately, nothing is simple. There are thousands of variables influencing the way programs are built, ranging from time of day to availability of remote servers and locale settings. Ensuring that programs are built the same way every time is surprising difficult, but it's a great way of establishing a chain of trust in which you, the end user, can be certain the program you're running is the same program the developer compiled, and that it doesn't have any malicious additional content added at some point in distribution.

andrewdavidwong commented:
Thanks, @marmarta! I think this intro is an improvement. @marmarek, if you're okay with adding it (modulo any changes you wish to make), let's do so. Then I can do a final editing pass before publication.

marmarek commented:
I've updated links, incorporated introduction by @marmarta and also added example for apt-transport-in-toto. I'll also describe progress on RPM side later today(?) (a lot has happened in the meantime).

marmarek commented:
I've added the missing part. @andrewdavidwong it is ready for final review now (and adjusting the date just before publishing).

andrewdavidwong commented:
@marmarek: I've pushed a revision. Since the revision is substantial in some areas, please review it to ensure you agree with the changes.

One remark: It stood out to me that the article never mentions making reproducible builds for Qubes ISOs, even as an ultimate long-term goal.

marmarek commented:
> One remark: It stood out to me that the article never mentions making reproducible builds for Qubes ISOs, even as an ultimate long-term goal.

I focused in this article on individual packages, as this is the part we are currently working on (and in fact I think it's more important than installation ISO - you easily build the ISO yourself if you don't trust official binaries, it is a one time work; but building each update separately is a much more tedious work). But indeed mentioning broader roadmap is a good idea.

marmarek commented:
Thanks @andrewdavidwong for the edits, looks good. I've added a final sentence as mentioned above. If that looks good to you, feel free to merge.
